src/colorizer/RecordingControls.ts

```diff
@@ -0,0 +1,147 @@
+    this.defaultPrefix = "";    
+    this.recording = false;
+    this.canvas = canvas;
+    this.timerId = 0;
+    this.startingFrame = 0;
+    this.redrawfn = redrawfn;
+    this.isDisabled = false;
+
+    this.startBtn = document.querySelector("#sequence_start_btn")!;
+    this.abortBtn = document.querySelector("#sequence_abort_btn")!;
+    this.filePrefixResetBtn = document.querySelector("#sequence_prefix_reset_btn")!;
+    this.startAtCurrentFrameChkbx = document.querySelector("#sequence_start_frame_checkbox")!;
+    this.filePrefixInput = document.querySelector("#sequence_prefix")!;
+
+    this.hiddenAnchorEl = document.createElement("a");  // Hidden element for initiating download later
+    document.body.appendChild(this.hiddenAnchorEl);
+
+    this.startBtn.addEventListener("click", () => this.handleStartButtonClick());
+    this.abortBtn.addEventListener("click", () => this.handleAbortButtonClick());
+    this.filePrefixInput.addEventListener("changed", () => this.handlePrefixChanged());
+    this.filePrefixResetBtn.addEventListener("click", () => this.handlePrefixResetClicked());
+  }
+
+  public setCanvas(canvas: ColorizeCanvas): void {
+    this.canvas = canvas;
+  }
+
+  private async handleStartButtonClick(): Promise<void> {
+    if (!this.recording) {
+      this.startingFrame = this.canvas.getCurrentFrame();
+      this.recording = true;
+
+      if (!this.startAtCurrentFrameChkbx.checked) {
+        await this.canvas.setFrame(0);  // start at beginning
+      }
+      this.startingFrame = this.canvas.getCurrentFrame();
+
+      this.startRecording();  // Start recording loop
+    }
+  }
+
+  private async handleAbortButtonClick(): Promise<void> {
+    if (this.recording) {
+      clearTimeout(this.timerId);
+      this.recording = false;
+      await this.canvas.setFrame(this.startingFrame);  // Reset to starting frame
+      this.redrawfn();
+    }
+  }
+
+  private async startRecording(): Promise<void> {
+    // Reset any existing timers
+    clearTimeout(this.timerId);
+
+    // Formatting setup
+    const maxDigits = this.canvas.getTotalFrames().toString().length;
+    
+    const loadAndRecordFrame = async (): Promise<void> => {
+      const currentFrame = this.canvas.getCurrentFrame();
+      
+      // Trigger a render through the redrawfn parameter so other UI elements update
+      this.redrawfn();
+      
+      // Get canvas as an image URL that can be downloaded
+      const dataURL = this.canvas.domElement.toDataURL("image/png");
+      const imageURL = dataURL.replace(/^data:image\/png/, "data:application/octet-stream");
+      
+      // Update our anchor (link) element with the image data, then force
+      // a click to initiate the download.
+      this.hiddenAnchorEl.href = imageURL;
+      const frameNumber: string = currentFrame.toString().padStart(maxDigits, "0");
+      this.hiddenAnchorEl.download = `${this.filePrefixInput.value}${frameNumber}.png`;
+      this.hiddenAnchorEl.click();
```

Comment:
I wonder if creating a completely new anchor element each time would also have helped the dropped frames?

Reply:
Tested this! It ran quite a bit slower than reusing the same anchor element (about 0:48) and dropped 18% of frames. This is comparable to the other runs-- I think the overhead of creating and adding some 570 anchor elements slows down the download loop and makes it comparable to the timeout.